### PR TITLE
v1.12.1 GPU and cleanup

### PR DESCRIPTION
**Pull request by danpetry**
Allows building of GPU variant
Also cleans up the recipe significantly, including:
  - unvendoring of as many libraries as possible
  - extensive commenting
  - enables distributed support for osx
  - removes any unused/mysterious legacy cruft

[Upstream](https://github.com/pytorch/pytorch)
[Upstream config env variables](https://github.com/pytorch/pytorch/blob/master/setup.py)
[Jira CPU variant](https://anaconda.atlassian.net/browse/PKG-360?atlOrigin=eyJpIjoiYWRjZDlkYzZlYTE0NDVhZDg1NDk1NWM5MTM0YmY2YmMiLCJwIjoiaiJ9)
[Jira GPU variant](https://anaconda.atlassian.net/browse/PKG-68?atlOrigin=eyJpIjoiNTQ3OTI4YzEzZWQwNDJhNDk2OTM1ZmViODdkOThlZGQiLCJwIjoiaiJ9)

# Status and testing

- GPU tarballs have been built out and tested and are on the atx-linuxgpu machines
- The build status of this PR for CPU is a little unclear from the github GUI because of the difficulty of building this package (it takes 24 hours). However all platforms build, because:
   - [this](https://github.com/AnacondaRecipes/pytorch-feedstock/pull/9/commits/f2d6ffe00600a3dd9d32d4ec1cfd6609e040f071) is the most recent commit that has been run through the CI. The two failures are platform failures. These platforms built successfully, see recent prefect logs [here](https://cloud.prefect.io/anaconda/flow-run/19f3833b-0e61-4bc1-9fdb-1235e646c54a) and [here](https://cloud.prefect.io/anaconda/flow-run/e419e3ea-73a5-4e5a-98e7-ad4afbaff80a).
   - changes after these flows are not relevant, i.e. wouldn't affect building
   - I also added a test for distributed support, but this should pass from inspecting the package
   - Even given the above, I should probably do another CI run to make sure it's good. But since this takes a day, it would be good to get some approving reviews while this is happening.

### CODE REVIEW

**Review by skupr-anaconda — COMMENTED** [review]
LGTM but folks more familiar with Pytorch GPU will be needed for the approval

File: recipe/meta.yaml [hunk 1]
```diff
@@ -114,20 +124,25 @@ requirements:
...
-    - libuv                           # [win]
+    - libprotobuf                     # [not win]
+    # on osx, libuv supports torch.distributed support. See build.sh.
+    - libuv                           # [win or osx]
     # Pinnings as per pytorch/.circleci/docker/common/install_conda.sh,
     # except for python 3.8 because we don't have the numpy version required
-    - numpy >=1.21.2,<1.22.0 # [py>=310]
-    - numpy >=1.19.2,<1.20.0 # [py<=39]
+    - numpy >=1.21,<1.22 # [py>=310]
+    - numpy >=1.19,<1.20 # [py<=39]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml:133]: Will it be more clear?
    - numpy 1.21  # [py>=310]
    - numpy 1.19  # [py<=39]

File: recipe/meta.yaml [hunk 2]
```diff
@@ -138,18 +153,21 @@ requirements:
...
     # other requirements
     - cffi
     # CF: pip check may fail if future is not installed
     - future
     # CF: needed to load C++ extensions
     - ninja
-    - numpy >=1.21.2,<2 # [py>=310]
-    - numpy >=1.19.2,<2 # [py<=39]
+    - numpy >=1.21,<2 # [py>=310]
+    - numpy >=1.19,<2 # [py<=39]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml:165]: There should be two spaces before `#`

File: recipe/meta.yaml (hunk 1 above)
Comment (danpetry) [reply recipe/meta.yaml:133]: This would resolve to 1.21.0 and 1.19.0

Comment (danpetry) [reply recipe/meta.yaml:133]: 1.21.* and 1.19.* is also possible

File: recipe/meta.yaml (hunk 2 above)
Comment (katietz) [reply recipe/meta.yaml:165]: true, please adjust that

File: recipe/meta.yaml (hunk 1 above)
Comment (katietz) [reply recipe/meta.yaml:133]: right ...

File: recipe/conda_build_config.yaml [hunk 3]
```diff
@@ -1,16 +1,36 @@
...
-# - gpu
+################# For GPU building, uncomment everything below and comment out the two lines above this #############
+## We currently haven't implemented and tested GPU acceleration on OSX (although it's possible for osx-arm64).
+## This is also the case for linux powerPC and ARM cores.
+## We also don't support it on Windows, because we need magma as a dependency and on windows, CMake isn’t playing ball
+## with building it with CUDA support.
+## So, only linux-64 for now.
+## Additionally, we need to pin the compiler if we're building with CUDA support (but not otherwise), see here:
+## https://docs.nvidia.com/cuda/archive/11.3.0/cuda-installation-guide-linux/index.html
+#pytorch_variant:       # [linux64]
```
Comment (katietz) [inline recipe/conda_build_config.yaml:27]: please replace all occurance of 'linux64' by '(linux and x86_64)'

File: recipe/README [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
-    CUDA 9.2: pkg_build_cos6_cuda92
-    CUDA 10.0: pkg_build_cos6_cuda100
 
-* Start the docker container using:
+# GPU building
 
-    ```
-    sudo nvidia-docker run -v `pwd`:/io -it pkg_build_cos6_cuda90
-    ```
+To build with GPU support, uncomment the line `#  - gpu                # [linux64 or win]`
```
Comment (katietz) [inline recipe/README:14]: use '(linux and x86_64)' instead of 'linux64' please

File: recipe/meta.yaml [hunk 5]
```diff
@@ -41,25 +30,35 @@ source:
...
     - patches/suppress_Wbitwise_instead_of_logical_clang.patch
+    # This is required in order to use our Pybind instead of the one
+    # submoduled into the Pytorch repo
+    - patches/cmake-fix-pybind-option.patch
+    # This patch was taken from conda-forge. There was a variable
+    # which directed the build system to use mkldnn for caffe2
+    # specifically, but since it was just turned on by the global
+    # mkldnn variable anyway, it was superfluous.
+    # Some of the changes in the original patch had been upstreamed.
+    - patches/cmake-unify-USE_MKLDNN-and-CAFFE2_USE_MKLDNN.patch
```
Comment (katietz) [inline recipe/meta.yaml:45]: please make sure that patches are treated always with linux line-endings.  See here .gitattributes file and adjust rules for *.diff and *.patch as needed

**Review by katietz — APPROVED** [review]
ok with those mentioned changes

File: recipe/meta.yaml (hunk 1 above)
Comment (danpetry) [reply recipe/meta.yaml:133]: I've left it as it is

File: recipe/meta.yaml (hunk 2 above)
Comment (danpetry) [reply recipe/meta.yaml:165]: done

File: recipe/conda_build_config.yaml (hunk 3 above)
Comment (danpetry) [reply recipe/conda_build_config.yaml:27]: done, do you mind if I ask why for my own understanding?

File: recipe/README (hunk 4 above)
Comment (danpetry) [reply recipe/README:14]: done

File: recipe/meta.yaml (hunk 5 above)
Comment (danpetry) [reply recipe/meta.yaml:45]: thanks, they are

**Review by varlackc — COMMENTED** [review]
The package seems to fail currently for `linux-ppc64le` We might need to see if we still want to provide support for that platform in this context, or if we want to skip the build.

**Comment by danpetry** [comment]
@varlackc it's a platform failure. Looking at the logs, we see that nothing ran. Others have been having similar issues.

**Review by skupr-anaconda — APPROVED** [review]

**Comment by danpetry** [comment]
osx-arm64 also can't build it. I'll start these builds manually.

**Review by sumit0190 — APPROVED** [review]
LGTM, great job and thanks!

**Review by varlackc — APPROVED** [review]
LGTM

**Pull request merged**